unstructured/ingest/connector/outlook.py

```diff
@@ -0,0 +1,214 @@
+    def get_file(self):
+        """Relies on Office365 python sdk message object to do the download."""
+        try:
+            if not self.download_dir.is_dir():
+                logger.debug(f"Creating directory: {self.download_dir}")
+                self.download_dir.mkdir(parents=True, exist_ok=True)
+
+            with open(
+                os.path.join(
+                    self.download_dir,
+                    self.hash_mail_name(self.file.id) + ".eml",
+                ),
+                "wb",
+            ) as local_file:
+                self.file.download(
+                    local_file,
+                ).execute_query()  # download MIME representation of a message
+
+        except Exception as e:
+            logger.error(
+                f"Error while downloading and saving file: {self.file.subject}.",
+            )
+            logger.error(e)
+            return
+        logger.info(f"File downloaded: {self.file.subject}")
+        return
+
+
+class OutlookConnector(ConnectorCleanupMixin, BaseConnector):
+    config: SimpleOutlookConfig
+
+    def __init__(
+        self,
+        standard_config: StandardConnectorConfig,
+        config: SimpleOutlookConfig,
+    ):
+        super().__init__(standard_config, config)
+        self._set_client()
+        self.get_folder_ids()
+
+    @requires_dependencies(["office365"])
+    def _set_client(self):
+        from office365.graph_client import GraphClient
+
+        self.client = GraphClient(self.config.token_factory)
+
+    def initialize(self):
+        pass
+
+    def recurse_folders(self, folder_id, main_folder_dict):
+        """We only get a count of subfolders for any folder.
+        Have to make additional calls to get subfolder ids."""
+        subfolders = (
+            self.client.users[self.config.user_email]
+            .mail_folders[folder_id]
+            .child_folders.get()
+            .execute_query()
+        )
+        for subfolder in subfolders:
+            for k, v in main_folder_dict.items():
+                if subfolder.get_property("parentFolderId") in v:
+                    v.append(subfolder.id)
+            if subfolder.get_property("childFolderCount") > 0:
+                self.recurse_folders(subfolder.id, main_folder_dict)
+
+    def get_folder_ids(self):
+        """Sets the mail folder ids and subfolder ids for requested root mail folders."""
+        self.root_folders = defaultdict(list)
+        root_folders_with_subfolders = []
+        get_root_folders = (
+            self.client.users[self.config.user_email].mail_folders.get().execute_query()
+        )
+
+        for folder in get_root_folders:
+            self.root_folders[folder.display_name].append(folder.id)
+            if folder.get_property("childFolderCount") > 0:
+                root_folders_with_subfolders.append(folder.id)
+
+        for folder in root_folders_with_subfolders:
+            self.recurse_folders(folder, self.root_folders)
```

Comment:
do we have folders and subfolders in our example? or does everything get flattened? probably okay if so since the file names are by id anyways, but want to make sure we're testing the nesting logic.

Reply:
We do have folders and subfolders. And I have tested double folders, etc...

Everything is returned flattened for simplicity. The way the api lists parent folders was less than ideal, so this was my simple solution. 

I'd be happy to look at subfolders in the partitioned return if you think it's important.

Reply:
Nope, I think that's fine. As long as getting the filename is ultimately unique and deterministic, since we currently rely on that to determine if we should process a file on subsequent runs.